openhtf/__init__.py

```diff
@@ -159,21 +164,27 @@ def add_output_callbacks(self, *callbacks):
     self._test_options.output_callbacks.extend(callbacks)
 
   def configure(self, **kwargs):
-    """Update test-wide configuration options.
-
-    Valid kwargs:
-      output_callbacks: List of output callbacks to run, typically it's better
-          to use add_output_callbacks(), but you can pass [] here to reset them.
-      teardown_function: Function to run at teardown.  We pass the same
-          arguments to it as a phase.
-    """
+    """Update test-wide configuration options. See TestOptions for docs."""
     # These internally ensure they are safe to call multiple times with no weird
     # side effects.
     create_arg_parser(add_help=True).parse_known_args()
     logs.setup_logger()
     for key, value in kwargs.iteritems():
       setattr(self._test_options, key, value)
 
+    # 'capture_source' takes effect here.
+    if self._test_options.capture_source:
+      # First, we copy the phases with the real CodeInfo for them.
+      phases = [
+        mutablerecords.CopyRecord(
+            phase, code_info=test_record.CodeInfo.for_function(phase.func))
+        for phase in self._test_desc.phases]
+
+      # Then we replace the TestDescriptor with one that stores the test
+      # module's CodeInfo as well as our newly copied phases.
+      self._test_desc = self._test_desc._replace(
+          code_info=self.code_info, phases=phases)
```

Comment:
After looking at it some more, storing it in an unconfigured state isn't possible since we call `configure()` inside of `__init__()` anyway. I'd rather not separate out `CodeInfo` since it's pretty relevant.

I think our options are instead:

* Make `capture_source` a special `metadata` attribute rather than a `TestOptions` attribute, which sort of makes sense since `metadata` is relevant to the `TestDescriptor` while `TestOptions` is relevant to a particular test setup.
* Make `capture_source` a config key instead and require it to be set before `Test` is created.
* A few more that I didn't like but I just thought of the config key and I think it's the best one so I'm going to do that